public/app/plugins/panel/timeseries/plugins/AnnotationEditorPlugin.tsx

```diff
@@ -0,0 +1,145 @@
+import React, { useLayoutEffect, useState, useCallback } from 'react';
+
+import { UPlotConfigBuilder, PlotSelection, usePlotContext } from '@grafana/ui';
+import { CartesianCoords2D, DataFrame, TimeZone } from '@grafana/data';
+import { AnnotationEditor } from './annotations/AnnotationEditor';
+
+type StartAnnotatingFn = (props: {
+  // pixel coordinates of the clicked point on the uPlot canvas
+  coords: { viewport: CartesianCoords2D; plotCanvas: CartesianCoords2D } | null;
+}) => void;
+
+interface AnnotationEditorPluginProps {
+  data: DataFrame;
+  timeZone: TimeZone;
+  config: UPlotConfigBuilder;
+  children?: (props: { startAnnotating: StartAnnotatingFn }) => React.ReactNode;
+}
+
+/**
+ * @alpha
+ */
+export const AnnotationEditorPlugin: React.FC<AnnotationEditorPluginProps> = ({ data, timeZone, config, children }) => {
+  const plotCtx = usePlotContext();
+  const [isAddingAnnotation, setIsAddingAnnotation] = useState(false);
+  const [selection, setSelection] = useState<PlotSelection | null>(null);
+
+  const clearSelection = useCallback(() => {
+    setSelection(null);
+    const plotInstance = plotCtx.plot;
+    if (plotInstance) {
+      plotInstance.setSelect({ top: 0, left: 0, width: 0, height: 0 });
+    }
+  }, [setSelection, plotCtx]);
+
+  useLayoutEffect(() => {
+    let annotating = false;
+
+    const setSelect = (u: uPlot) => {
+      if (annotating) {
+        setIsAddingAnnotation(true);
+        const min = u.posToVal(u.select.left, 'x');
+        const max = u.posToVal(u.select.left + u.select.width, 'x');
+
+        setSelection({
+          min,
+          max,
+          bbox: {
+            left: u.select.left,
+            top: 0,
+            height: u.bbox.height / window.devicePixelRatio,
+            width: u.select.width,
+          },
+        });
+        annotating = false;
+      }
+    };
+
+    config.setCursor({
+      bind: {
+        mousedown: (u, targ, handler) => (e) => {
+          if (e.button === 0) {
+            handler(e);
+            if (e.metaKey) {
+              annotating = true;
+            }
+          }
+
+          return null;
+        },
+        mouseup: (u, targ, handler) => {
+          return (e) => {
+            if (e.button === 0) {
+              if (annotating) {
+                let _setSelectHooks;
+
+                // Monkey patch existing setSelect hooks (ZoomPlugin for instance)
```

Comment:
oof, i should see if there's a more elegant way to API this :grimacing:

Reply:
yes, this was my major concern- but it isn't wors than patching setSelect in your demo :)

Reply:
it feels a bit worse, since that doesnt swap a core object, just toggles a boolean.

can we do something like this in the other setSelect hooks instead?

```js
if (panelCtx.isAnnotating)
  return;
```

Reply:
another, maybe better way to do this is for AnnotationEditorPlugin to add an `init` hook and use that to wrap each of the setSelect hooks in a function that uses the above logic with a local `isAnnotating` check.

Reply:
✅